[Quant][PT2E]Fix pt2e quantization maxpool input observer issue

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98826
* #98730
* __->__ #100961

**Summary**
Fix the issue https://github.com/pytorch/pytorch/issues/100959. The root cause is for node of `torch.ops.aten.max_pool2d_with_indices.default`, there are 2 output node as output tensor and max indices. So in its `node.meta["val"]` is a tuple of `FakeTensors` (For example: `'val': (FakeTensor(..., size=(1, 2, s1, s1)), FakeTensor(..., size=(1, 2, s1, s1), dtype=torch.int64))`). It will fail the check  of inserting observer since which only accept one `FakeTensor` case. 

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

leslie-fang-intel committed:
Fix pt2e quantization maxpool input observer issue

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jerryzh168, May I kindly ask a question about this piece of logic?https://github.com/pytorch/pytorch/blob/f08ddae8885bc703408b949642d4a5bee30efce8/torch/ao/quantization/_pt2e/prepare.py#L151-L155
Why we skip of inserting observer if the output is not a tensor? Does the change in this PR looks good to you?

jerryzh168 commented:
why do we change this? I don't think we want to insert observers for non-Tensor values

jerryzh168 commented:
oh I see, I think the skip logic is not correct, we should only skip inserting output observer in this case

jerryzh168 commented on torch/ao/quantization/_pt2e/prepare.py:
```diff
@@ -22,9 +22,14 @@ def _maybe_insert_input_and_output_observers_for_node(
         output_is_a_tensor = (
             this_node_dtype_info is not None and
-            isinstance(node.meta["val"], FakeTensor)
+            check_node_meta_val_fake_tensor(node.meta["val"])
```
not this one, L41 should be applied after L56

leslie-fang-intel replied:
Looks like we can't simply move L41 after L56 due to the case https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/_pt2e/prepare.py#L30
When the node has no `target_dtype_info`, we should also skip the insert of input observer. Does my understanding right?

leslie-fang-intel replied:
Move L41 after L56 and also add `skip_inserting_input_observers` to skip the insert of input observer when `node.meta["target_dtype_info"]` not set for this node. @jerryzh168 Please help to take a look of this PR again.

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix pt2e quantization maxpool input observer issue"


**Summary**
Fix the issue https://github.com/pytorch/pytorch/issues/100959. The root cause is for node of `torch.ops.aten.max_pool2d_with_indices.default`, there are 2 output node as output tensor and max indices. So in its `node.meta["val"]` is a tuple of `FakeTensors` (For example: `'val': (FakeTensor(..., size=(1, 2, s1, s1)), FakeTensor(..., size=(1, 2, s1, s1), dtype=torch.int64))`). It will fail the check  of inserting observer since which only accept one `FakeTensor` case. 

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix pt2e quantization maxpool input observer issue"


**Summary**
Fix the issue https://github.com/pytorch/pytorch/issues/100959. The root cause is for node of `torch.ops.aten.max_pool2d_with_indices.default`, there are 2 output node as output tensor and max indices. So in its `node.meta["val"]` is a tuple of `FakeTensors` (For example: `'val': (FakeTensor(..., size=(1, 2, s1, s1)), FakeTensor(..., size=(1, 2, s1, s1), dtype=torch.int64))`). It will fail the check  of inserting observer since which only accept one `FakeTensor` case. 

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix pt2e quantization maxpool input observer issue"


**Summary**
Fix the issue https://github.com/pytorch/pytorch/issues/100959. The root cause is for node of `torch.ops.aten.max_pool2d_with_indices.default`, there are 2 output node as output tensor and max indices. So in its `node.meta["val"]` is a tuple of `FakeTensors` (For example: `'val': (FakeTensor(..., size=(1, 2, s1, s1)), FakeTensor(..., size=(1, 2, s1, s1), dtype=torch.int64))`). It will fail the check  of inserting observer since which only accept one `FakeTensor` case. 

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/_pt2e/prepare.py:
```diff
@@ -29,11 +29,11 @@ def _maybe_insert_input_and_output_observers_for_node(
 
-    skip_inserting_observers = (
-        not output_is_a_tensor
+    skip_inserting_input_observers = (
```
nit: this should probably be `skip_inserting_input_and_output_observers`

leslie-fang-intel replied:
Changed and thanks.

jerryzh168 reviewed (approved):
LGTM, we are landing a refactor diff right now though: https://github.com/pytorch/pytorch/pull/101041 maybe you can land after that is merged

leslie-fang-intel commented:
It's for 

> LGTM, we are landing a refactor diff right now though: #101041 maybe you can land after that is merged

Sure, will land this PR after that.

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix pt2e quantization maxpool input observer issue"


**Summary**
Fix the issue https://github.com/pytorch/pytorch/issues/100959. The root cause is for node of `torch.ops.aten.max_pool2d_with_indices.default`, there are 2 output node as output tensor and max indices. So in its `node.meta["val"]` is a tuple of `FakeTensors` (For example: `'val': (FakeTensor(..., size=(1, 2, s1, s1)), FakeTensor(..., size=(1, 2, s1, s1), dtype=torch.int64))`). It will fail the check  of inserting observer since which only accept one `FakeTensor` case. 

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Fix pt2e quantization maxpool input observer issue"


**Summary**
Fix the issue https://github.com/pytorch/pytorch/issues/100959. The root cause is for node of `torch.ops.aten.max_pool2d_with_indices.default`, there are 2 output node as output tensor and max indices. So in its `node.meta["val"]` is a tuple of `FakeTensors` (For example: `'val': (FakeTensor(..., size=(1, 2, s1, s1)), FakeTensor(..., size=(1, 2, s1, s1), dtype=torch.int64))`). It will fail the check  of inserting observer since which only accept one `FakeTensor` case. 

**Test Plan**
```
python -m pytest test_quantize_pt2e.py -k test_max_pool2d_quantizer
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
